### PR TITLE
Make kubelet set alpha.kubernetes.io/provided-node-ip unconditionally

### LINKED ISSUE #109793: Enabling external CCM causes Node.Addresses to flap
### What happened?

I am reproducing this in OpenShift using the OpenStack cloud provider, but nothing here is specific to either OpenShift or OpenStack.

When migrating a cluster which uses in-tree cloud provider to use external cloud-provider, we use the following process:

1. Disable cloud controllers in KCM
2. Deploy CCM
3. Redeploy kubelets with --cloud-provider=external

Between steps 1&2 the node-lifecycle, routes, and service controllers are not running anywhere, but this situation isn't unsafe for the cluster and we can keep it short.

Between steps 2&3 is more interesting. The node controller is running in CCM, and the kubelets are running with --cloud-provider=openstack (or whatever, just not external). The main loop of node-controller will ignore a newly created node because [it won't have the external cloud provider taint](https://github.com/kubernetes/kubernetes/blob/1234d061c0bdc53b41bc7c12881bbaa7af084e55/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L394-L399). However, the [UpdateNodeStatus()](https://github.com/kubernetes/kubernetes/blob/1234d061c0bdc53b41bc7c12881bbaa7af084e55/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L166) loop does not have a similar exclusion.

[Kubelet also manages Node.Addresses](https://github.com/kubernetes/kubernetes/blob/1234d061c0bdc53b41bc7c12881bbaa7af084e55/pkg/kubelet/nodestatus/setters.go#L62-L209). Note that we shortcut if the kubelet is using an external cloud provider, but not if we don't. So between steps 2&3 both kubelet and CCM's node controller are managing Node.Addresses. If the legacy cloud provider and external CCM don't return the same list of NodeAddresses this causes flapping.

Running OpenShift on OpenStack we have 2 problems. Firstly, our legacy cloud provider and external CCM produce different lists. We don't think this would be a problem in practise, except for the second problem.

Kubelet adds the `alpha.kubernetes.io/provided-node-ip` annotation to the node containing the node IP. This allows the CCM node controller to apply the same filtering based on node IP as kubelet does for the legacy cloud provider. This means that, although the lists are different, the node IP is still listed first which is the only thing that matters in practise.

The problem is that kubelet only adds the `alpha.kubernetes.io/provided-node-ip` annotation when --cloud-provider=external which, between steps 2&3, it is not. This means that Node.Addresses flaps between a correct and a broken configuration, as the NodeAddresses returned by the external CCM cannot be filtered by node IP without the annotation.

The ideal solution here would be to ensure that only 1 controller is managing Node.Addresses during this period in the same way we do for Node creation with the external CCM taint. This would fix the issue entirely.

A fix with a much smaller scope would be to have kubelet add the node IP annotation unconditionally so, although Node.Address can potentially flap, it will at least be between 2 valid configurations. Given that this should only ever be a problem during the upgrade window, the smaller fix may be more expedient.

### What did you expect to happen?

Node.Addresses should remain valid at all times when kubelet is configured with --cloud-provider=openstack but the CCM is also running.

### How can we reproduce it (as minimally and precisely as possible)?

Good luck 😂

### Anything else we need to know?

_No response_

### Kubernetes version

1.24

### Cloud provider

OpenStack (but as described, not OpenStack-specific)

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by mdbooth**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
The easier of two possible fixes for an issue where node addresses flap during an upgrade to an external CCM. This fix causes kubelet to apply the alpha.kubernetes.io/provided-node-ip annotation unconditionally (not only when --cloud-provider=external). This does not exclude a future fix involving the root cause of the issue, which is that kubelet and cloud-controller-manager both attempt to manage node addresses when --cloud-provider is not set to external.

#### Which issue(s) this PR fixes:

Fixes #109793

#### Does this PR introduce a user-facing change?

```release-note
The node annotation alpha.kubernetes.io/provided-node-ip is no longer set ONLY when `--cloud-provider=external`.  Now, it is set on kubelet startup if the `--cloud-provider` flag is set at all, including the deprecated in-tree providers.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/nodestatus/setters.go [hunk 1]
```diff
@@ -94,14 +94,21 @@ func NodeAddress(nodeIPs []net.IP, // typically Kubelet.nodeIPs
...
-					node.ObjectMeta.Annotations = make(map[string]string)
-				}
-				node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr] = nodeIP.String()
+		// Annotate the Node object with nodeIP for external cloud
+		// provider. We do this even when external CCM is not configured
+		// to cover a situation during upgrade when CCM is running the
+		// node controller in the cluster but kubelet is still running
+		// the in-tree provider. Adding this annotation in all cases
+		// ensures that while Addresses flap between the competing
+		// controllers, they at least flap consistently.
```
Comment (elmiko) [inline pkg/kubelet/nodestatus/setters.go:103]: nice comment, thanks Matt

**Review by elmiko — COMMENTED** [review]
i am not super familiar with code here, but this generally makes sense to me
/lgtm

**Comment by nckturner** [comment]
cc @kubernetes/sig-node-pr-reviews

**Comment by nckturner** [comment]
ping @yujuhong @kubernetes/sig-node-bugs @kubernetes/sig-node-pr-reviews, who is the correct reviewer/approver from sig-node for this?

**Comment by dchen1107** [comment]
Is this a regression from 1.23? Do we need to cherrypick?

**Comment by nckturner** [comment]
@dchen1107 It's not a regression, but it is a bug.  I think we should definitely cherry pick.

**Comment by aojea** [comment]
/assign @danwinship @andrewsykim 

hmm, changing the behavior to ALWAYS annotate just to solve one corner and specific case seems excessive to me 
Why clusters without any cloud provider at all has to carry with a cloud provider annotation, that is also alpha by the way, and this will make it harder to move in the future

**Comment by danwinship** [comment]
> Why clusters without any cloud provider at all has to carry with a cloud provider annotation

The patch would work if it only annotated in clusters where `cloudProvider != ""` (ie, external or legacy cloud provider, but not bare-metal).

> that is also alpha by the way, and this will make it harder to move in the future

It's more like a GA annotation with "alpha" in its name :slightly_frowning_face:. Though yeah, using it more would make it harder to fix that.

**Comment by aojea** [comment]
> The patch would work if it only annotated in clusters where `cloudProvider != ""` (ie, external or legacy cloud provider, but not bare-metal).

ah, ok, excuse me then

**Comment by danwinship** [comment]
I didn't check the code. The fix may not literally be `cloudProvider != ""` but that _idea_ should work

**Comment by mdbooth** [comment]
I have updated the PR with the request from @aojea to make no change for the non cloud-provider case.

I have also added unit tests which verify this behaviour.

**Comment by mdbooth** [comment]
Failures both look like flakes

/retest-required

**Comment by nckturner** [comment]
Tests passed @mdbooth @aojea

**Review by elmiko — COMMENTED** [review]
i think this is looking good, thanks for the updates @mdbooth 
/lgtm

**Comment by aojea** [comment]
> > ```
> > if externalCloudProvider || cloud != nil {
> >      do annotation
> > } 
> > ```
> 
> I'll make this change.

thanks for your comprehension, defer approval for people more familiar with cloud-provider codebase

/assign @cheftako @thockin @andrewsykim

**Comment by markyjackson-taulia** [comment]
Hi @mdbooth  My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

**Comment by mdbooth** [comment]
> Hi @mdbooth  My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

I certainly hope so! Nothing to do from my side that I'm aware of beyond soliciting approvals.

**Comment by dchen1107** [comment]
Another PR I came across might try to address a similar issue caused by the external cloud provider migration: https://github.com/kubernetes/kubernetes/pull/110704 cc @haiqianz

File: pkg/kubelet/nodestatus/setters.go [hunk 2]
```diff
@@ -94,14 +94,28 @@ func NodeAddress(nodeIPs []net.IP, // typically Kubelet.nodeIPs
 			klog.V(4).InfoS("Using secondary node IP", "IP", secondaryNodeIP.String())
 		}
 
-		if externalCloudProvider {
+		if externalCloudProvider || cloud != nil {
+			// Annotate the Node object with nodeIP for external cloud provider.
```
Comment (dchen1107) [inline pkg/kubelet/nodestatus/setters.go:98]: You are changing the semantic of the pre-defined annotation: https://kubernetes.io/docs/reference/labels-annotations-taints/#alpha-kubernetes-io-provided-node-ip. Change the related doc?

**Review by dchen1107 — COMMENTED** [review]
One nit on the related doc. Otherwise LGTM

**Comment by nckturner** [comment]
/lgtm

Who can approve?  @dchen1107?

**Comment by aojea** [comment]
I just found there is a KEP to handle this in/out tree cloud provider migration problem https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2436-controller-manager-leader-migration , can not be used for this?

**Comment by mdbooth** [comment]
> I just found there is a KEP to handle this in/out tree cloud provider migration problem https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2436-controller-manager-leader-migration , can not be used for this?

Unfortunately not. We're effectively migrating the 'Node controller' from running locally in every kubelet to running centrally in a single Node controller. We'd need a separate leader election for each individual kubelet.

If we wanted to add more annotations I suppose the kubelet could annotate the Node to indicate who is expected to manage it, and the Node controller could be taught to ignore kubelet-managed Nodes. However, this annotation would have an extremely limited useful lifetime.

**Comment by nckturner** [comment]
This PR involves the kubelet and the CCM, (the above KEP refers to a lock between the KCM and CCM) and is specific to the period of time when migration is taking place (old kubelets are still running and reconciling Node Addresses, and the new CCM also begins reconciling node addresses).  In order for the above leader migration lock to be used for this, kubelet would need to take that migration lock before doing its node address management. Might be possible, but I think regardless we should still make the above change, because they are not mutually exclusive and it is a simple remediation to the flapping address behavior that is described above.

**Comment by dchen1107** [comment]
@nckturner thanks for open pr for doc updates: https://github.com/kubernetes/website/pull/35027

/lgtm
/approve

**Comment by aojea** [comment]
> @nckturner I recommend adding a changelog entry (not `NONE`) for this change. The updated behavior is visible to end users.

agreed

**Comment by aojea** [comment]
/release-note-edit ```release-note The node annotation alpha.kubernetes.io/provided-node-ip is no longer set ONLY when `--cloud-provider=external`.  Now, it is set on kubelet startup if the `--cloud-provider` flag is set at all, including the deprecated in-tree providers. ```

**Pull request merged**